// action-docs/lib/cli.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require(".");
const chalk_1 = __importDefault(require("chalk"));
const figlet_1 = __importDefault(require("figlet"));
const linebreak_1 = require("./linebreak");
const yargs_1 = __importDefault(require("yargs"));
const args = yargs_1.default.options({
    "toc-level": {
        description: "TOC level used for markdown",
        type: "number",
        default: _1.defaultOptions.tocLevel,
        demandOption: false,
        alias: "t",
    },
    action: {
        description: "GitHub action file",
        type: "string",
        default: _1.defaultOptions.actionFile,
        demandOption: false,
        alias: "a",
    },
    "no-banner": {
        description: "Print no banner",
        requiresArg: false,
    },
    "update-readme": {
        description: "Update readme file.",
        requiresArg: false,
        type: "string",
        alias: "u",
    },
    "line-breaks": {
        description: "Used line breaks in the generated docs.",
        default: "LF",
        choices: ["CR", "LF", "CRLF"],
        demandOption: false,
        type: "string",
        alias: "l",
    },
}).argv;
args["banner"] === undefined &&
    console.info(chalk_1.default.blue(figlet_1.default.textSync("ACTION-DOCS", { horizontalLayout: "full" })));
const updateReadme = args["update-readme"] === undefined ? false : true;
/* eslint-disable github/no-then */
(0, _1.generateActionMarkdownDocs)({
    actionFile: args.action,
    tocLevel: args["toc-level"],
    updateReadme,
    readmeFile: args["update-readme"] === undefined || args["update-readme"] === ""
        ? _1.defaultOptions.readmeFile
        : args["update-readme"],
    lineBreaks: (0, linebreak_1.getLineBreakType)(args["line-breaks"]),
})
    .then((r) => {
    if (!updateReadme) {
        console.info(r);
    }
})
    .catch((e) => console.error(e.message));
/* eslint-enable */
